### ISSUE TITLE
Negation operators don't work as expected for meta queries

### ISSUE THREAD

**Issue by e-dard** [influxdata/influxdb#9976]
The negation operators (`!=` and `!~`) don't work as expected for some meta queries. The following meta queries are not affected:

```
SHOW SERIES
SHOW TAG VALUES
SHOW FIELD KEYS
```

Here are examples of affected meta queries:

#### Example 1 - SHOW MEASUREMENTS

This example fails on `tip` and also on `1.4.2` and `1.3.6`.
```
> create database db
> use db
Using database db
>
> insert cpu,region=west value=1
> insert cpu,region=east value=1
>
> show measurements
name: measurements
name
----
cpu
>
> show measurements where region = 'west'
name: measurements
name
----
cpu
>
> show measurements where region = 'east'
name: measurements
name
----
cpu
>
> show measurements where region != 'east'
>
> show measurements where region != 'west'
>
```

In both of these cases we should expect the `cpu` measurement to be returned. The issue here is that because there exist a series in the measurement that doesn't satisfy `!= 'east'`, the entire measurement is excluded from the result set. To clarify this a little further:

```
> create database db
> use db
Using database db
>
> insert cpu,region=east value=1
> insert cpu,region=west value=1
> insert mem,region=east value=1
>
> show measurements where region = 'east'
name: measurements
name
----
cpu
mem
>
> show measurements where region != 'west'
name: measurements
name
----
mem
>
>
```
The `!= 'west'` returns `mem` because `mem` does not have a series where `region = west`. 

#### Example 2 - SHOW TAG KEYS

```
> drop database db
> create database db
>
> insert cpu,region=west value=1
> insert cpu,region=east value=1
>
> show tag keys
name: cpu
tagKey
------
region
>
> show tag keys where region != 'east'
> show tag keys where region !~ 'west'
ERR: error parsing query: found west, expected regex at line 1, char 30
> show tag keys where region !~ /west/
```

The same problem as `SHOW MEASUREMENTS`. The negation options will exclude any tag key if there is an example series where the predicate is not true.

The issue is to do with the way the [tags are matched](https://github.com/influxdata/influxdb/blob/master/tsdb/index.go#L1310-L1434) when looking for measurements to return.

**Comment by benbjohnson** [comment]
@pauldix This would be a breaking change for the query language. Can you weigh in on this? Should this be implemented or should we defer to Flux?

tl;dr—meta queries like `SHOW MEASUREMENTS WHERE region != 'west'` will current not return a measurement if any of its series contains `region=west`. The proposal is to return the measurement if it contains any series where `region` is not `'west'`.

**Comment by pauldix** [comment]
Yeah, I think because it's a breaking change we shouldn't do anything for the 1.x line. For context, this came up in the discussion of how negation works in Flux. Take for example the following series:

```
_measurement=foo,region=west
_measurement=foo,region=east
_measurement=bar,region=west
_measurement=asdf,region=east
```

If you have the Flux query

```
from() |> range() |> filter(fn: (r) => r.region != "west")
```

You'd still have a series in the result set with `_measurement=foo` because one series matches. Representing the other kind of logic would be a bit more tricky.

cc @nathanielc

**Comment by e-dard** [comment]
@pauldix this is the same underlying issue we discussed with regards to the prometheus read endpoint too. It's an index issue, which means it affects 1.x query engine, and anything else that uses the index such as the `storage` service (prometheus read / flux).

**Comment by pauldix** [comment]
@e-dard so we need to update the index to handle it, but really only for the storage service, not for InfluxQL
